Mount kubelet and container runtime rootdir on LSSD

alculquicondor opened the pull request:
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

When environment variable NODE_LOCAL_SSD_EPHEMERAL=true,
create a RAID 0 array on all attached Local SSDs on NVMe interfaces to mount:

- kubelet root dir
- container runtime root dir
- pod logs dir

Those directories account for all ephemeral storage.

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

alculquicondor commented:
/assign @mattcary @msau42

msau42 commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+    echo "y" | mdadm --create "${md_device}" --level=0 --raid-devices=${#devices[@]} ${devices[@]}
+  fi
+  local ephemeral_mountpoint="/mnt/disks/kube-ephemeral-ssd"
+  safe-format-and-mount "${md_device}" "${ephemeral_mountpoint}"
```
Are the mount options and directory permissions what we want, such as the "a+w" permissions?

alculquicondor replied:
yes, verified.

msau42 commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+  # Some images mount the container runtime root dir.
+  umount "/var/lib/${container_runtime}" || true
+  mkdir -p "${ephemeral_mountpoint}/${container_runtime}"
+  safe-bind-mount "${ephemeral_mountpoint}/${container_runtime}" "/var/lib/${container_runtime}"
```
Same question here wrt permissions and mount options

alculquicondor replied:
Yes, verified.

msau42 commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -376,6 +376,10 @@ function mount-ext(){
 # Local ssds, if present, are mounted or symlinked to their appropriate
 # locations
 function ensure-local-ssds() {
+  if [ "${NODE_LOCAL_SSDS_EPHEMERAL:-false}" == "true" ]; then
```
Do we need more plumbing of this env? Like in configure-default.sh, configure-test.sh?

alculquicondor replied:
What are the effects or those?

msau42 replied:
Those are used for kube-up.sh and kubetest

alculquicondor replied:
Done

alculquicondor commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+    if [ -e "${ssd}" ]; then
+      # This workaround to find if the NVMe device is a disk is required because
+      # the existing Google images does not expose NVMe devices in /dev/disk/by-id
+      if [[ `udevadm info --query=property --name=${ssd} | grep DEVTYPE | sed "s/DEVTYPE=//"` == "disk" ]]; then
```
It was copied over from `ensure-local-ssds`

msau42 replied:
I think the existing logic doesn't handle PD as nvme and assumes all nvme devices are local SSDs. @mattcary do you want to restrict this new ephemeral option to only scsi PD boot disk for now, and update this logic to handle nvme PD boot disk later?

alculquicondor replied:
We want nvme support. Is there any doc on how to detect nvme PDs?

alculquicondor replied:
using ID_MODEL=nvme_card

alculquicondor replied:
Updated to lsblk

alculquicondor commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+  local devices=(/dev/disk/by-id/google-local-ssd-*)
+  if [ ! -e ${devices[0]} ]; then
+    echo "No local SCSI SSD disks found."
+    devices=()
```
It's kind of the intention, but should we enforce it here?

mattcary replied:
It seems like the conservative choice would be to restrict to one or the other. Allowing both maybe has the possibility of the node getting wedged in a weird way. Being restrictive on how we do things might be more reliable.

alculquicondor replied:
Any preference on which should take preference if both happen to be present? Or I could just remove the checks for scsi.

mattcary replied:
Preferring NVMe seems right.

If this happens, does it mean that someone has done some weird manual configuration? So the fact that the ephemeral storage is less than the total attache LSSD won't be a surprise.

alculquicondor replied:
Yeah, it would only be due to manual overrides.

alculquicondor replied:
Removed scsi support

alculquicondor commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+
+  # mount container runtime root dir on SSD
+  local container_runtime="${CONTAINER_RUNTIME:-docker}"
+  systemctl stop ${container_runtime}
```
Fixed.

alculquicondor commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
   done
 }
 
+function ensure-local-ssds-ephemeral-storage() {
```
Done

alculquicondor commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+    md_device="/dev/md/0"
+    echo "y" | mdadm --create "${md_device}" --level=0 --raid-devices=${#devices[@]} ${devices[@]}
+  fi
+  local ephemeral_mountpoint="/mnt/disks/kube-ephemeral-ssd"
```
The rest of /mnt is read-only. I see this for `/mnt/disks`:

```
tmpfs /mnt/disks tmpfs rw,relatime,size=256k,mode=755 0 0
```

One possibility would be `/mnt/stateful_partition/ephemeral_storage`, but then we are putting it inside the boot disk.

alculquicondor replied:
Done

mattcary commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -425,6 +429,56 @@ function ensure-local-ssds() {
+    fi
+  done
+  if [ "${#devices[@]}" -eq 0 ]; then
+    return
```
Can we error or log or something here? I don't know how we can surface node startup problems, but if there is one it would be nice if it were possible to see it clearly in logs or whatever.

alculquicondor replied:
We are already logging in the lines above. If there are no SSDs, I don't think it should be considered an error.

mattcary replied:
ok, I agree those logs should be enough.

alculquicondor commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -406,25 +410,83 @@ function ensure-local-ssds() {
+      if [[ `udevadm info --query=property --name=${ssd} | grep ID_MODEL | sed "s/ID_MODEL=//"` != "nvme_card" ]]; then
+        continue
       fi
+      local devicenum=`echo ${ssd} | sed -e 's/\/dev\/nvme0n\([0-9]*\)/\1/'`
```
I think we can't rely on it being `nvme0n`, but we can fix in a followup

mattcary replied:
Yeah let's not change existing behavior (which is what ensure-local-ssds is for, right?)

mattcary replied:
(although nvme[0-9]* would probably do the trick?)

alculquicondor replied:
assuming that doesn't produce a match group. I need to refresh my `sed` knowledge

alculquicondor replied:
I reverted changes in the existing function. Let's fix that in a followup

alculquicondor committed:
Mount kubelet and container runtime rootdir on LSSD

When environment variable NODE_LOCAL_SSD_EPHEMERAL=true,
create a RAID 0 array on all attached SSDs to mount:

- kubelet root dir
- container runtime root dir
- pod logs dir

Those directories account for all ephemeral storage.
An array is not created when there is only one SSD.

Change-Id: I22137f1d83fc19e9ef58a556d7461da43e4ab9bd
Signed-off-by: Aldo Culquicondor <acondor@google.com>

msau42 commented:
/lgtm
/assign @cheftako

mattcary commented:
/lgtm
as well

alculquicondor commented:
/retest

alculquicondor commented:
Friendly ping @cheftako 

Giving the freeze, we are not looking for merging it now, but just to get early feedback

cheftako commented on cluster/gce/util.sh:
```diff
@@ -1215,6 +1215,7 @@ CONTAINER_RUNTIME_TEST_HANDLER: $(yaml-quote ${CONTAINER_RUNTIME_TEST_HANDLER:-}
 UBUNTU_INSTALL_CONTAINERD_VERSION: $(yaml-quote ${UBUNTU_INSTALL_CONTAINERD_VERSION:-})
 UBUNTU_INSTALL_RUNC_VERSION: $(yaml-quote ${UBUNTU_INSTALL_RUNC_VERSION:-})
 NODE_LOCAL_SSDS_EXT: $(yaml-quote ${NODE_LOCAL_SSDS_EXT:-})
+NODE_LOCAL_SSDS_EPHEMERAL: $(yaml-quote ${NODE_LOCAL_SSDS_EPHEMERAL:-})
```
Thanks :D

cheftako commented:
/lgtm

alculquicondor commented:
/retest

alculquicondor commented:
Turns out the tool complains about the creation of a single device array.

I'm skipping the array creation for this case. PTAL

alculquicondor commented:
/retest

msau42 commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -449,15 +449,19 @@ function ensure-local-ssds-ephemeral-storage() {
-    echo "y" | mdadm --create "${md_device}" --level=0 --raid-devices=${#devices[@]} ${devices[@]}
+  local device="${devices[0]}"
+  if [ "${#devices[@]}" -ne 1 ]; then
+    local device=(/dev/md/*)
```
Is this going to set a new variable locally, and not modify the one at L452?

alculquicondor replied:
Oddly enough, it worked in manual testing. Fixed.

msau42 commented:
/lgtm

alculquicondor commented:
/retest

alculquicondor commented:
/hold

I'm checking some kube-proxy issues

alculquicondor commented:
/hold cancel

Fixed problems with preloaded images. PTAL

cheftako commented:
/lgtm

alculquicondor commented:
/retest

msau42 commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -479,7 +479,11 @@ function ensure-local-ssds-ephemeral-storage() {
+  # Move the container runtime's directory to the new location to preserve
+  # preloaded images.
+  if [ ! -d "${ephemeral_mountpoint}/${container_runtime}" ]; then
+    mv "/var/lib/${container_runtime}" "${ephemeral_mountpoint}/${container_runtime}"
```
We unmounted "/var/lib/${container_runtime}" above. Is there any data here to move?

alculquicondor replied:
They are rw remounts. I have yet to verify in Ubuntu, though. I'll get back to this.

mattcary replied:
Sorry, I don't follow. Do you mean that the cases where there are preloaded images, /var/lib/docker isn't mounted so the unmount is a no-op?

alculquicondor replied:
I rechecked. The OS images that mount are the containerd ones. And it's just a RW remount in the same disk. We need to unmount prior to moving the folder.

This is independent of having preloaded images or not.

mattcary replied:
I'm still not following. Why do we need to unmount? If we're concerned about someone writing to it during the move, the umount isn't enough because it will silently fail if the dir is already in use (ie there's a race).

alculquicondor replied:
We need to unmount to be able to move the data. Otherwise we would get a "resource busy" error. Nothing is writing to it, as the container runtime is stopped.

mattcary replied:
finally I understand, bind-mounted dirs make that resource busy error. thx

alculquicondor commented:
There are new shell check requirements. PTAL at last 2 commits

mattcary commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -436,6 +440,62 @@ function ensure-local-ssds() {
+    seen_arrays=(/dev/md/*)
+    device=${seen_arrays[0]}
+    echo "Setting RAID array with local SSDs on device ${device}"
+    if [ ! -e "$device" ]; then
```
Why can we assume an existing raid array has to be our local SSDs?

mattcary replied:
answer: we have to, because it's too complicated to figure out what an existing raid is from when the node is restarted.

alculquicondor replied:
There's no other mechanism that creates RAID arrays in the startup script.

mattcary commented on cluster/gce/gci/configure-helper.sh:
```diff
@@ -436,6 +440,62 @@ function ensure-local-ssds() {
+  # mount container runtime root dir on SSD
+  local container_runtime="${CONTAINER_RUNTIME:-docker}"
+  systemctl stop "$container_runtime"
+  # Some images mount the container runtime root dir.
```
Would it be more precise to say "remount" here? That gives a better hint as to what's going on IMHO.

alculquicondor replied:
Done

mattcary commented:
/lgtm

fwiw

alculquicondor commented:
Rebased and squashed

@cheftako

alculquicondor commented:
/retest

mattcary commented:
Still 
/lgtm

alculquicondor commented:
ping @cheftako 

/retest

cheftako commented:
/lgtm
/approve
